src/main/kotlin/com/zoltu/gradle/plugin/GitVersioning.kt

```diff
@@ -64,25 +72,59 @@ class GitVersioning : Plugin<Project> {
 	}
 
 	private fun setProjectVersion(project: Project, versionInfo: VersionInfo) {
-		project.version = versionInfo.toString()
+		project.version = configuration.customVersionToString.apply(versionInfo)
 	}
 
 	private fun setJarManifestVersion(project: Project, versionInfo: VersionInfo) {
 		val closure = object {
 			@Suppress("unused")
 			fun apply(jar: Jar) {
 				val attributes = jar.manifest?.attributes ?: throw Exception("The Jar task has no manifest.")
-				attributes.put("Implementation-Version", versionInfo.toString())
+				attributes.put("Implementation-Version", configuration.customVersionToString.apply(versionInfo))
 				attributes.put("Implementation-Sha", versionInfo.sha)
 				attributes.put("Specification-Version", "${versionInfo.major}.${versionInfo.minor}")
 			}
 		}
 		(project.tasks ?: throw Exception("The project has no tasks.")).withType(Jar::class.java, MethodClosure(closure, "apply"))
 	}
 
-	private fun exposeVersionInfoToUser(project: Project, versionInfo: VersionInfo) {
-		project.extensions.create("ZoltuGitVersioning", Extension::class.java, versionInfo)
-	}
+	private fun exposeConfigurationObject(project: Project, configurationChangeCallback: () -> Unit) = project.extensions.create("ZoltuGitVersioning", Configuration::class.java, configurationChangeCallback)
 
-	open class Extension(open val versionInfo: VersionInfo)
+	open class Configuration(private val configurationChangeCallback: () -> Unit) {
+		companion object {
+			@Suppress("unused")
+			const val NAME = "versioning"
+			private val defaultVersionToString = Function<VersionInfo, String> { versionInfo: VersionInfo ->
+				val suffix = if (versionInfo.tags != null && versionInfo.patch != null) {
+					// semantic versioning
+					"${versionInfo.patch}-${versionInfo.tags}.${versionInfo.commitCount}"
+				} else if (versionInfo.patch != null) {
+					// semantic versioning
+					"${versionInfo.patch}-${versionInfo.commitCount}"
+				} else if (versionInfo.tags != null) {
+					// simple versioning
+					"${versionInfo.commitCount}-${versionInfo.tags}"
+				} else {
+					// simple versioning
+					"${versionInfo.commitCount}"
+				}
+				"${versionInfo.major}.${versionInfo.minor}.$suffix".toString()
+			}
+			private val defaultCustomDescribeProcessor = Function<String, VersionInfo?> { null }
+		}
+		/**
+		 * Contains version information for consumption.
+		 */
+		lateinit var versionInfo: VersionInfo
+
+		/**
+		 * A custom describe processor that will take in a string as input and return a VersionInfo object as output.
+		 */
+		var customDescribeProcessor: Function<String, VersionInfo?> by observable(defaultCustomDescribeProcessor, { property, old, new -> configurationChangeCallback() })
```

Comment:
I actually started with a regex provider, as you had.  The problem there was that it required the user to have a regex with the right number of captures in the right places.  As I was writing the doc-comment and error handling for it, I realized that it was complex to explain to end users and *really* prone to errors, which is why I fell-back to the factory mechanism instead.

I wonder if there is some way to get strong typing on the `VersionInfo` object in IntelliJ, ideally with auto-complete?  I wonder if Kotlin gradle files have the same problem or not?

Another option would be to have the factory return an array of Strings?  I don't like this because instead of failing fast, you fail slow and debugging is more difficult but presumably logic could be added in the plugin to give very clear messages here.  This will "solve" the IntelliJ problem, but at a cost.

Another option would be to use a RegEx with named captures rather than ordered captures.  I haven't tried this in Java, is it a thing one can do (my RegEx-foo is weak)?